src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java

```diff
@@ -444,13 +457,371 @@ protected boolean checkCgIcrRules(String subAccountTypeCode, SubAccountGlobalDet
+    }
+    
+    protected void putFieldErrorForApplyAllCheckboxRulesFailure(String propertyName, String propertyDisplayName, String errorConstant) {
+        String checkboxDisplayName = getDisplayNameForGlobalObjectProperty(CUKFSPropertyConstants.APPLY_TO_ALL_NEW_SUB_ACCOUNTS);
+        putFieldError(propertyName, errorConstant, new String[] {propertyDisplayName, checkboxDisplayName});
+    }
+    
+    protected boolean checkDuplicateOrExistingSubAccountsAreNotPresent(SubAccountGlobal subAccountGlobal) {
+        String commonSubAccountNumber = subAccountGlobal.getNewSubAccountNumber();
+        String subAccountKeyFormat = "%s-%s-%s";
+        Set<String> subAccountKeys = new HashSet<>();
+        boolean success = true;
+        
+        int i = 0;
+        for (SubAccountGlobalNewAccountDetail newAccountDetail: subAccountGlobal.getSubAccountGlobalNewAccountDetails()) {
+            String subAccountNumber = subAccountGlobal.isApplyToAllNewSubAccounts()
+                    ? commonSubAccountNumber : newAccountDetail.getSubAccountNumber();
+            String subAccountKey = String.format(subAccountKeyFormat, newAccountDetail.getChartOfAccountsCode(),
+                    newAccountDetail.getAccountNumber(), subAccountNumber);
+            if (!subAccountKeys.add(subAccountKey)) {
+                putFieldErrorForDuplicateOrExistingSubAccount(
+                        CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_NEW_SUB_ACCOUNT_DUPLICATE, subAccountKey, i);
+                success = false;
+            } else {
+                success &= checkSubAccountDoesNotExist(newAccountDetail, subAccountNumber, subAccountKey, i);
+            }
+            i++;
+        }
+        
+        return success;
+    }
+    
+    protected boolean checkSubAccountDoesNotExist(
+            SubAccountGlobalNewAccountDetail newAccountDetail, String subAccountNumber, String subAccountKey, int index) {
+        SubAccount subAccount = getSubAccountService().getByPrimaryId(
+                newAccountDetail.getChartOfAccountsCode(), newAccountDetail.getAccountNumber(), subAccountNumber);
+        if (ObjectUtils.isNotNull(subAccount)) {
+            putFieldErrorForDuplicateOrExistingSubAccount(
+                    CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_NEW_SUB_ACCOUNT_EXISTS, subAccountKey, index);
+            return false;
+        }
+        return true;
+    }
+    
+    protected void putFieldErrorForDuplicateOrExistingSubAccount(String errorConstant, String subAccountKey, int index) {
+        String fullPropertyName = buildListObjectPropertyPath(
+                CUKFSPropertyConstants.SUB_ACCOUNT_GLOBAL_NEW_ACCOUNT_DETAILS, KFSPropertyConstants.SUB_ACCOUNT_NUMBER, index);
+        putFieldError(fullPropertyName, errorConstant, subAccountKey);
+    }
+    
+    protected boolean checkContractsAndGrantsSetupForNewAccounts(SubAccountGlobal subAccountGlobal) {
+        boolean success = true;
+        boolean allCG = true;
```

Comment:
I think this is a good flow to determine if all the accounts are the same in terms of needing CG detail or not.  However, this code reads a bit confusing, can this be refactored for read-ability?

Reply:
Ok, I'll see if I can make this method more readable.